### PR TITLE
HPL1: add engine and game code

### PR DESCRIPTION
**Pull request by grisenti**

### CODE REVIEW

File: graphics/tinygl/init.cpp [hunk 1]
```diff
@@ -261,7 +261,7 @@ void GLContext::init(int screenW, int screenH, Graphics::PixelFormat pixelFormat
 	// color mask
 	color_mask_red = color_mask_green = color_mask_blue = color_mask_alpha = true;
 
-	const int kDrawCallMemory = 5 * 1024 * 1024;
```
Comment (aquadran) [inline graphics/tinygl/init.cpp:264]: that allocation is too big, it's 1GB RAM! try find matching value

File: engines/hpl1/credits.pl [hunk 2]
```diff
@@ -0,0 +1,3 @@
+begin_section("Hpl1");
```
Comment (aquadran) [inline engines/hpl1/credits.pl:1]: we need fill entries here. @sev-  what you could suggest to put here?

File: engines/hpl1/configure.engine [hunk 3]
```diff
@@ -0,0 +1,3 @@
+# This file is included from the main "configure" script
+# add_engine [name] [desc] [build-by-default] [subengines] [base games] [deps]
+add_engine hpl1 "Hpl1" yes
```
[the remark this answers is not in the document]
Comment (aquadran) [reply engines/hpl1/configure.engine:3]: and engine should be disabled by default, as it's not completable yet

File: image/codecs/codec.cpp [hunk 4]
```diff
@@ -205,7 +205,7 @@ Codec *createBitmapCodec(uint32 tag, uint32 streamTag, int width, int height, in
 
 	switch (tag) {
 	case SWAP_CONSTANT_32(0):
-		return new BitmapRawDecoder(width, height, bitsPerPixel, true);
```
[the remark this answers is not in the document]
Comment (sev-) [reply image/codecs/codec.cpp:208]: Could you please elaborate a bit, because I did this change

File: engines/hpl1/credits.pl (hunk 2 above)
Comment (sev-) [reply engines/hpl1/credits.pl:1]: Emanuele Grisenti, perhaps @digitall (David Turner). My contribution is not significant. also, perhaps @aquadran (Paweł Kolodziejski)

Comment (digitall) [reply engines/hpl1/credits.pl:1]: My contributions to HPL1 inclusion are mainly minor fixes for warnings, though I do have a copy of the game so looking forward to playing reliably (the original is twitchy about OpenGL drivers / cards resulting in glitches when I tried to play it on Win32).

Comment (digitall) [reply engines/hpl1/credits.pl:1]: If you want to add me, feel free, but up to @grisenti who has done most of the heavy work here :)

File: graphics/tinygl/init.cpp (hunk 1 above)
Comment (grisenti) [reply graphics/tinygl/init.cpp:264]: I had forgotten about this. Testing a complex scene it looks like 60mb is the maximum of what is needed, though it's usually less (around 20mb)

File: image/codecs/codec.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (aquadran) [reply image/codecs/codec.cpp:208]: last time I looked into, it was about BMP pixel format, BMP can have ARGB or XRGB, that can be detected based on header version

File: engines/hpl1/console.cpp [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "hpl1/console.h"
```
Comment (sev-) [inline engines/hpl1/console.cpp:22]: I think it is safe to remove this file.

File: engines/hpl1/console.cpp [hunk 6]
```diff
@@ -0,0 +1,38 @@
+/* ScummVM - Graphic Adventure Engine
```
Comment (sev-) [inline engines/hpl1/console.cpp:1]: I think it is safe to remove this file. It has no actual content anyway besides the boilerplate code.

File: engines/hpl1/console.h [hunk 7]
```diff
@@ -0,0 +1,41 @@
+
```
Comment (sev-) [inline engines/hpl1/console.h:1]: This file is not needed as well.

File: engines/hpl1/debug.h [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+	kDebugTextures = 1 << 7
+};
+
+enum DebugLevels {
+	kDebugLevelError = 1,
+	kDebugLevelWarning = 2,
+	kDebugLevelLog = 3
+};
+
+template<typename... Args>
```
Comment (sev-) [inline engines/hpl1/debug.h:48]: I would recommend just mass-replace `log*` with `debug*`

File: engines/hpl1/detection.cpp [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+};
+
+Hpl1MetaEngineDetection::Hpl1MetaEngineDetection() : AdvancedMetaEngineDetection(Hpl1::GAME_DESCRIPTIONS,
+	sizeof(ADGameDescription), Hpl1::GAME_NAMES) {
+}
+
+DetectedGames Hpl1MetaEngineDetection::detectGames(const Common::FSList &fslist, uint32 skipADFlags, bool skipIncomplete) {
+	DetectedGames games; 
+	for(auto file : fslist) {
+		if (file.getName() == "Penumbra.exe")
```
Comment (sev-) [inline engines/hpl1/detection.cpp:52]: `file.getName.compareIgnoreCase("Penumbra.exe")`

File: engines/hpl1/detection.cpp [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+	DEBUG_CHANNEL_END
+};
+
+Hpl1MetaEngineDetection::Hpl1MetaEngineDetection() : AdvancedMetaEngineDetection(Hpl1::GAME_DESCRIPTIONS,
+	sizeof(ADGameDescription), Hpl1::GAME_NAMES) {
+}
+
+DetectedGames Hpl1MetaEngineDetection::detectGames(const Common::FSList &fslist, uint32 skipADFlags, bool skipIncomplete) {
+	DetectedGames games; 
+	for(auto file : fslist) {
```
Comment (sev-) [inline engines/hpl1/detection.cpp:51]: Formatting

File: engines/hpl1/detection_tables.h [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+namespace Hpl1 {
+
+const PlainGameDescriptor GAME_NAMES[] = {
+	{"penumbraoverture", "Penumbra: Overture"},
+	{0, 0}};
+
+const ADGameDescription GAME_DESCRIPTIONS[] = {
+	{"penumbraoverture",
+	 nullptr,
+	 AD_ENTRY1s("Penumbra.exe", nullptr, -1),
```
Comment (sev-) [inline engines/hpl1/detection_tables.h:31]: It is time to add real data here

File: engines/hpl1/std/map.h [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#ifndef HPL1_STD_MAP_H
+#define HPL1_STD_MAP_H
+
+#include "hpl1/algorithms.h"
+#include "pair.h"
```
Comment (sev-) [inline engines/hpl1/std/map.h:26]: Please use full paths here starting from "hpl1/"

File: engines/hpl1/std/map.h [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+	}
+
+	Common::Array<value_type> _items;
+	CompFunc _comp;
+};
+
+} // namespace std
+} // namespace Hpl1
+
+#endif
```
Comment (sev-) [inline engines/hpl1/std/map.h:191]: No end line

File: engines/hpl1/std/map.h [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+ */
+
+#ifndef HPL1_STD_MAP_H
+#define HPL1_STD_MAP_H
+
+#include "hpl1/algorithms.h"
+#include "pair.h"
+
+namespace Hpl1 {
+namespace std {
```
Comment (sev-) [inline engines/hpl1/std/map.h:29]: I'd highly recommend renaming this to `Std` for all files.

**Review by sev- — COMMENTED** [review]
Added couple more notes

**Comment by sev-** [comment]
Merged

File: dists/scummvm.rc [hunk 15]
```diff
@@ -196,6 +196,82 @@ shaders/playground3d_cube.vertex        FILE    "engines/playground3d/shaders/pl
 shaders/playground3d_fade.fragment      FILE    "engines/playground3d/shaders/playground3d_fade.fragment"
 shaders/playground3d_fade.vertex        FILE    "engines/playground3d/shaders/playground3d_fade.vertex"
 #endif
+#if PLUGIN_ENABLED_STATIC(HPL1)
+shaders/hpl1_Ambient_Color.fragment                    FILE   "engines/hpl1/engine/impl/shaders/hpl1_Ambient_Color.fragment"
```
[the remark this answers is not in the document]
Comment (aquadran) [reply dists/scummvm.rc:200]: that approach require some rework on prioritise which files to load first. when you are running from sources, the files from sources have priority over *.dat files. also some script which will combine files into dat file

File: Makefile.common [hunk 16]
```diff
@@ -508,6 +508,9 @@ endif
 ifdef ENABLE_PLAYGROUND3D
 DIST_FILES_SHADERS+=$(wildcard $(srcdir)/engines/playground3d/shaders/*)
 endif
+ifdef ENABLE_HPL1
+DIST_FILES_SHADERS+=$(wildcard $(srcdir)/engines/hpl1/engine/impl/shaders/*)
```
[the remark this answers is not in the document]
Comment (aquadran) [reply Makefile.common:514]: there is no problem with that. but like above

File: image/codecs/codec.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (aquadran) [reply image/codecs/codec.cpp:208]: I don't see this addressed, @sev- we have now regression on trunk, assuming previous was not broken

**Pull request closed without merging**